bootloader/grub.go

```diff
@@ -132,14 +138,173 @@ func (g *grub) SetBootVars(values map[string]string) error {
 	return env.Save()
 }
 
+func (g *grub) extractedKernelDir(prefix string, s snap.PlaceInfo) string {
+	return filepath.Join(
+		prefix,
+		filepath.Base(s.MountFile()),
+	)
+}
+
 func (g *grub) ExtractKernelAssets(s snap.PlaceInfo, snapf snap.Container) error {
-	// XXX: should we use "kernel.yaml" for this?
-	if _, err := snapf.ReadFile("meta/force-kernel-extraction"); err == nil {
-		return extractKernelAssetsToBootDir(g.dir(), s, snapf)
+	// default kernel assets are:
+	// - kernel.img
+	// - initrd.img
+	// - dtbs/*
+	var assets []string
+	if g.uefiRunKernelExtraction {
+		assets = []string{"kernel.efi"}
+	} else {
+		assets = []string{"kernel.img", "initrd.img", "dtbs/*"}
+	}
+
+	// extraction can be forced through either a special file in the kernel snap
+	// or through an option in the bootloader
+	_, err := snapf.ReadFile("meta/force-kernel-extraction")
+	if g.uefiRunKernelExtraction || err == nil {
+		return extractKernelAssetsToBootDir(
+			g.extractedKernelDir(g.dir(), s),
+			s,
+			snapf,
+			assets,
+		)
 	}
 	return nil
 }
 
 func (g *grub) RemoveKernelAssets(s snap.PlaceInfo) error {
 	return removeKernelAssetsFromBootDir(g.dir(), s)
 }
+
+// ExtractedRunKernelImageBootloader helper methods
+
+func (g *grub) makeKernelEfiSymlink(s snap.PlaceInfo, name string) error {
+	// don't use g.dir() for the symlink destination ("oldname" in go parlance)
+	// because we want a relative symlink target, i.e. we want a symlink like:
+	//
+	// kernel.efi -> pc-kernel_1.snap/kernel.efi
+	//
+	// always, even if grub is really working from
+	// /run/mnt/ubuntu-boot/EFI/ubuntu or at runtime from /boot/grub where the
+	// symlink is like:
+	//
+	// /boot/grub/kernel.efi -> /boot/grub/pc-kernel_1.snap/kernel.efi
+	//
+	extractedKernel := filepath.Join(
+		filepath.Base(s.MountFile()),
+		"kernel.efi",
+	)
+
+	// check that the kernel snap has been extracted already so we don't
+	// inadvertently create a dangling symlink here
+	// here we do add rootdir since we're following where the symlink will end
+	// up
+	if !osutil.FileExists(filepath.Join(g.dir(), extractedKernel)) {
+		return fmt.Errorf(
+			"cannot enable %s at %s: %v",
+			name,
+			extractedKernel,
+			os.ErrNotExist,
+		)
+	}
+	return os.Symlink(
+		extractedKernel,
+		// use g.dir() for where we are creating the symlink
+		filepath.Join(g.dir(), name),
+	)
+}
+
+func (g *grub) unlinkKernelEfiSymlink(name string) error {
+	symlink := filepath.Join(g.dir(), name)
+	if osutil.FileExists(symlink) {
+		return osutil.UnlinkMany(g.dir(), []string{name})
```

Comment:
Why `osutil.UnlinkMany`? Is it because `os.Remove()` doesn't call `unlinkat`?

Reply:
It was more of a convenience because we already had the relative filename and dir, but it's simpler to understand if I just use os.Remove, so I refactored to use that.